Don't emit to_time deprecations in known-safe contexts

matthewd opened the pull request:
I realised that #51994 was producing a deprecation warning while running the Active Record tests (I actually thought we'd made that fail CI, but I must be mis-remembering).

While we could just explicitly opt in to the new setting within our test suite(s), I wanted to see if we could save similar callers [i.e. those that aren't particularly concerned with time zones] from having to know about / act upon the change at all.


Mostly that comes down to the fact that for `Time#to_time`, if the value is already a local time, there's no difference in behaviour, and so no need to warn.

Correspondingly, I've avoided calling `to_time` in the handful of places we were using it internally: it's easy to do, and we know we don't care about the zone, so again won't be affected by the change.

I'm hoping this will mean that people who do see the deprecation are likely to actually be affected [or at least be genuinely at risk of being affected] by the change, thus justifying the disruption.

It's worth noting that I currently have _not_ changed `ActiveSupport::TimeWithZone#to_time`: even when its zone matches local (including the most common case where both are UTC), we will still warn. That felt like the right choice at the time, though I'm now less sure... `3.days.ago.to_time` seems like a pretty reasonable and ordinary thing to be doing, and (in the absence of any zone-related configuration) I believe it should be perfectly safe. :thinking:

matthewd committed:
Don't emit to_time deprecations in known-safe contexts

If the value is already a local time, there's no difference, so no need
to warn.

Correspondingly, avoid calling to_time in the handful of places we were
using it internally: it's easy to do, and we know we don't care about
the zone.

matthewd commented:
> `3.days.ago.to_time` seems like a pretty reasonable and ordinary thing to be doing, and (in the absence of any zone-related configuration) I believe it should be perfectly safe

I've now reminded myself that if you haven't configured a `Time.zone`, then `3.days.ago` will be a plain `Time`, not a TWZ, so this PR will already save you the deprecation.

I think that puts my thinking back to the current state of the PR: as long as you have (unremarkably zoned) Time instances, we can let you `to_time` without comment; if you've got a TWZ, we should continue to infer that zones are significant for your purposes, and warn.

jhawthorn commented on activesupport/lib/active_support/core_ext/time/calculations.rb:
```diff
@@ -319,7 +319,12 @@ def compare_with_coercion(other)
     elsif other.is_a?(Time)
-      compare_without_coercion(other.to_time)
+      # also avoid ActiveSupport::TimeWithZone#to_time before Rails 8.0
+      if other.respond_to?(:comparable_time)
+        compare_without_coercion(other.comparable_time)
+      else
+        compare_without_coercion(other.to_time)
+      end
```
(non-blocking) If we know this is a Time subclass (or is claiming to be one like TimeWithZone) how would you feel about just using `utc` (since that exists on time and is the same as `comparable_time`)?

```suggestion
      compare_without_coercion(other.utc)
```

jhawthorn committed:
Avoid using to_time in TimeWithZone#-

We're more likely to already have utc and can avoid constructing the
extra time object, and this will avoid deprecations on unconfigured
to_time_preserves_timezone.

jhawthorn commented:
@matthewd pushed a commit to your branch removing one more common and unnecessary `to_time`.
